ocs_ci/utility/psiutils.py

```diff
@@ -0,0 +1,151 @@
+            size (int): of the volume in GBs
+            volume_type (str): type of volume to be created
+            metadata (dict): Any {k,v} to be associated with volume
+
+        Returns:
+            Instance of "Volume" class
+
+        """
+        cluster_tag = {'cluster_name': config.ENV_DATA['cluster_name']}
+        meta_data = metadata if metadata else cluster_tag
+
+        vol = self.cinder_clnt.volumes.create(
+            name=name,
+            size=size,
+            volume_type=volume_type,
+            metadata=meta_data
+        )
+        if not vol:
+            raise exceptions.PSIVolumeCreationFailed(
+                "Failed to create PSI openstack volume"
+            )
+
+        return vol
+
+    def attach_volume(self, vol, instance_id, mnt_pnt):
+        """
+        Attach the given volume to the specific PSI openstack instance
+
+        Args:
+            vol_id (Volume): cinder volume object
+            instance_id (str): uuid of the instance
+            mnt_pnt (str): path where volume will be mounted
+
+        """
+        self.cinder_clnt.volumes.attach(
+            vol, instance_id, mnt_pnt
+        )
+
+    def get_instances_with_pattern(self, pattern):
+        """
+        Get instances matching pattern
+
+        Args:
+            pattern (str): Pattern for matching instance name
+                note: we are looking for only 'name' param for
+                pattern matching
+
+        Returns:
+            novaclient.base.ListWithMeta
+
+        """
+        return self.nova_clnt.servers.list(
+            search_opts={'name': pattern}
+        )
+
+    def get_volumes_with_tag(self, tag):
+        """
+        Get PSI volumes having this tag (k,v)
+
+        Args:
+            tag (dict): of desired (k,v)
+
+        Returns:
+            list: of cinderclient.<vers>.volumes object
+
+        """
+        matching_vols = []
+        for vol in self.cinder_clnt.volumes.list():
+            for k, v in tag.items():
+                if vol.metadata.get(k) == v:
+                    matching_vols.append(vol)
+
+        return matching_vols
+
+    def detach_and_delete_vols(self, vollist):
+        """
+        Detach and delete volumes from the list
+
+        Args:
+            list: of Volume objects
+
+        """
+        for v in vollist:
+            if v.status == 'in-use':
+                v.detach()
+                v.get()
+                if v.status == 'available':
+                    logger.info(f"Volume {v.name} detached successfully ")
+                else:
+                    for res in TimeoutSampler(60, 1, v.status == 'available'):
+                        if not res:
+                            logger.info(f"Waiting for {v.name} to detach")
+                            v.get()
+            v.delete()
+            try:
+                v.get()
+            except Exception:
+                logger.info(f"Volume {v.name} deleted successfully")
```

Comment:
It will be better to handle one particular exception there, this general `except Exception` might hide some other issues (e.g. API outage,...).
And what if the volume will not be deleted correctly or if it will some time?

Reply:
ack